### ISSUE TITLE
[BUG]: Modifying typed answer with JS works inconsistently

### ISSUE THREAD

**Issue by Eltaurus-Lt** [ankidroid/Anki-Android#16213]
### Checked for duplicates?

- [X] This issue is not a duplicate

### Does it also happen in the desktop version?

- [X] This bug does not occur in the latest version of Anki

### What are the steps to reproduce this bug?

1. Enable 'Type answer into the card' in the settings
2. add the following script to the Front side of a card in the `Basic (type in the answer)` Note Type:
```
<script>
document.getElementById('typeans').value = 'a';
</script>
```
3. During the review submit an answer by pressing the "Show answer" button

### Expected behaviour

The back of the card is expected to show, that the submitted answer is 'a'. (it does so on Anki desktop)

### Actual behaviour

The back of the card shows that an empty answer was submitted.

### Debug info

```text
AnkiDroid Version = 2.17.6 (568538896be943818ca793dd2635822265712def)

Backend Version = 0.1.34-anki23.12.1 (23.12.1 1a1d4d5419c6b57ef3baf99c9d2d9cf85d36ae0a)

Android Version = 11 (SDK 30)

ProductFlavor = play

Manufacturer = HMD Global

Model = Nokia C01 Plus

Hardware = T19545AA1

Webview User Agent = Mozilla/5.0 (Linux; Android 11; Nokia C01 Plus Build/RP1A.201005.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/123.0.6312.118 Mobile Safari/537.36

ACRA UUID = bb072dd9-2ecf-49e6-81a4-c0e99b4f0fd8

Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

If the Android keyboard is used to type anything else in the input field after it was modified with JS, the back of the card shows the full answer. E.g., if 'b' is manually typed after step 2 and before step 3, the submitted answer on the back of the card is correctly shown to be "ab".

It looks like AnkiDroid uses a separate variable to track the typed answer, instead of directly reading the value of html element, and this value is updated only when the on-screen keyboard is used. In that case, maybe it is possible to add the call of the updating function into the "Show answer" button as well to ensure that the submitted answer is always the same as what is displayed on the screen?

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by oyeraghib** [comment]
Hi @Eltaurus-Lt I would need your help with two things as I was not able to reproduce the above issue.

1. Can you provide a mobile app video with showing what you have shared in `(Optional) Anything else you want to share?` section. I was able to reproduce the issue where it doesn't show the `'a'` text on pressing `Show answer` but at the same time if user manually types `b`, I couldn't see it becoming `ab`. A video of that showing it would be really helpful.

2. Secondly I tried it on Anki Desktop, it seems to be working in the same way as mobile, is it possible to show a video of that too where it works fine on desktop. Also the steps to reproduce remains same on desktop ? (As I was not able to find Step 1 there).

Thanks.

**Comment by Eltaurus-Lt** [comment]
Sure
1. Here is the video for AnkiDroid:
![ankidroid](https://github.com/ankidroid/Anki-Android/assets/93875472/5f441e7e-17f1-4aed-bda6-a5a655cdb152)
It is recorded with Chrome remote debugging tools, so the interface is not visible. The sequence of buttons I press here during the review: `Show Answer` → `Again` → Type 'b' with the appearing Android on-screen keyboard → `Show Answer`. You can see that the typed answer is empty on the back of the card the first time (2 is the expected answer), while it is correctly displayed as "ab" the second time.
2.  >Also the steps to reproduce remains same on desktop ? (As I was not able to find Step 1 there).

   Step 1 is for AnkiDroid solely, as it is required to disable the default Android input in favor of using the card HTML input field, which is the only input method on desktop as is.
   Here is the video that shows the reviewing of the same card as above on Anki desktop as well as its full code:
![anki desktop](https://github.com/ankidroid/Anki-Android/assets/93875472/5c573f63-dc8d-4d7e-8493-3bd39170b4d2)
(my desktop version is "⁨23.12.1 (1a1d4d54)⁩ Python 3.9.15 Qt 6.6.1 PyQt 6.6.1" by the way)

**Comment by Eltaurus-Lt** [comment]
> 2\. (my desktop version is "⁨23.12.1 (1a1d4d54)⁩ Python 3.9.15 Qt 6.6.1 PyQt 6.6.1" by the way)

I also just installed "Version ⁨24.04.1 (ccd9ca1a)⁩ Python 3.9.18 Qt 6.6.2 PyQt 6.6.1" and tested the card there. It works exactly the same.

**Comment by Eltaurus-Lt** [comment]
still searching for solutions

**Comment by SlugFiller** [comment]
Hit the same issue. Instead of setting the value on page load, it's possible to set it on some event, and get a similar result, e.g.
```html
{{type:Answer}}
<a id="change_answer" href="#">Click here to change the answer</a>
<script type="text/javascript">
document.getElementById('change_answer').addEventListener('click', (e) => {
  e.preventDefault();
  document.getElementById('typeans').value = 'Changed value';
});
</script>
```
If you then type an answer, and then press the button, the back will show the typed answer, and not the set value.

P.S. Instead of messing with the element, it would be nice if the JS API had functions to read/write/listen for changes on the native edit text, so "Type answer into the card" is not necessary.

**Comment by Eltaurus-Lt** [comment]
Yes, I also arrived at the minimum example in the original post when [debugging an on-screen keyboard embedded into a card](https://forums.ankiweb.net/t/modifying-typeans-contents-with-js/43612).
The workaround was to make a custom `typeans` element:
```
<input id="typeans" type="text" inputmode="text" autocorrect="off" autocomplete="off" autocapitalize="off" spellcheck="false">
```
store its value in sessionStorage to read it on the back of a card, then add an event listener to track changes and update the current state at each key pressed.

**Comment by Eltaurus-Lt** [comment]
_still searching for solutions_

**Comment by SlugFiller** [comment]
still searching for solutions

**Comment by SlugFiller** [comment]
I do not know of any changes or updates made to resolve this situation. The fact that it's not being addressed isn't going to make it go away.

**Comment by Eltaurus-Lt** [comment]
still looking for solutions as well

**Comment by david-allison** [comment]
@Eltaurus-Lt Is this still an issue in the new study screen?

**Comment by davidbelius** [comment]
I am having this isuse on AnkiDroid 2.23.0alpha5, with the new study screen enabled.

There seems to be no element with ID "typeans" in the DOM that is accessible to javascript in a card <script> block.

**Comment by david-allison** [comment]
@BrayanDSO FYI

**Comment by Eltaurus-Lt** [comment]
> [@Eltaurus-Lt](https://github.com/Eltaurus-Lt) Is this still an issue in the new study screen?

(missed this one smh)
Can confirm that in the current v2.23.1 release with the new study screen enabled this is no longer an issue, both in the minimal example and in the original card type.
